### PR TITLE
Update infrastrucure to support storage jobs

### PR DESCRIPTION
**Pull request by markcowl**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Storage/Storage.sln [hunk 1]
```diff
@@ -34,6 +34,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Authorizati
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Graph.RBAC", "..\..\Common\Commands.Common.Graph.RBAC\Commands.Common.Graph.RBAC.csproj", "{269ACF73-0A34-42DC-AB9C-4B15931A489D}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Profile", "..\Profile\Commands.Profile\Commands.Profile.csproj", "{142D7B0B-388A-4CEB-A228-7F6D423C5C2E}"
```
Comment (cormacpayne) [inline src/ResourceManager/Storage/Storage.sln]: @markcowl why was this reference added?

File: src/Storage/Commands.Storage/Blob/Cmdlet/GetAzureStorageBlob.cs [hunk 2]
```diff
@@ -114,6 +114,9 @@ public string Container
         [Parameter(Mandatory = false, HelpMessage = "Continuation Token.")]
         public BlobContinuationToken ContinuationToken { get; set; }
 
+        [Parameter(Mandatory =false, HelpMessage ="Run this task in the background and return a job object for tracking execution progress.")]
+        public SwitchParameter AsJob { get; set; }
```
Comment (cormacpayne) [inline src/Storage/Commands.Storage/Blob/Cmdlet/GetAzureStorageBlob.cs]: @markcowl any test(s) needed for this change?

Comment (cormacpayne) [inline src/Storage/Commands.Storage/Blob/Cmdlet/GetAzureStorageBlob.cs]: @markcowl please regenerate the markdown help file for this cmdlet to add `AsJob`

File: src/Common/Commands.Common/AzureLongRunningJob.cs [hunk 3]
```diff
@@ -243,13 +243,22 @@ public override string StatusMessage
                 returnValue.MyInvocation.BoundParameters.Add(parameter.Key, parameter.Value);
             }
 
-            foreach (var field in returnType.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly))
+            CopyCmdletFields(returnType, cmdlet, returnValue);
+            cmdlet.SafeCopyParameterSet(returnValue);
+            return returnValue as U;
+        }
+
+        static void CopyCmdletFields<U>(Type targetType, U sourceCmdlet, U targetCmdlet) where U : AzurePSCmdlet
```
Comment (cormacpayne) [inline src/Common/Commands.Common/AzureLongRunningJob.cs]: @markcowl not sure if it's possible, but could an object of type `AzurePSCmdlet` be passed as `targetType` in this method? If so, is there any problem with copying the properties of the `PSCmdlet` type (which `AzurePSCmdlet` inherits from)?

**Comment by cormacpayne** [comment]
@markcowl also, just saw this was marked as `Do Not Merge`, so feel free to ignore my comments until this is ready ☺️

**Comment by blueww** [comment]
@markcowl 
I tried to apply the changes in the PR, and run Get-AzureStorageBlob and Set-AzureStorageBlobContent with -asjob, both still fail:

```
PS C:\windows\system32> $job = Set-AzureStorageBlobContent -Container 123 -File C:\temp\0.txt -Force -AsJob

PS C:\windows\system32> $job.Error
Value cannot be null.
Parameter name: path1
    + CategoryInfo          : InvalidOperation: (Microsoft.Azure....AzurePSCmdlet]:AzureLongRunningJob`1) [], ArgumentNullException
    + FullyQualifiedErrorId : The cmdlet failed in background execution.  The returned error was 'Value cannot be null.
Parameter name: path1'.  Please execute the cmdlet again.  You may need to execute this cmdlet synchronously, by omitting the '-AsJob' parameter.
```

**Pull request closed without merging**